File: Project 3/backtrack.java
/*
    Gabbi Forsythe
    12/12/2019
    CS351-01
    Dr. Chen

    
    PROJECT 3: Part 1 - 0/1 Knapsack (Backtracking Algorithm)
    Use the backtracking algorithm to find the solution to the 0/1 Knapsack problem
*/

import java.io.*;
import java.util.Scanner;

class backtrack {
    // Used global variables because that is how the book does it in Algorithm 5.7 (pg.242-243)
    static int numItems, capacity, maxProfit, numbest;
    static int [] w;
    static int [] p; 
    static boolean [] include;
    static boolean [] bestset;

    public static void knapsack (int i, int profit, int weight) {
        System.out.println("Knapsack - "+ i);
        if ((weight <= capacity) && (profit > maxProfit)) {     // if it is the best set so far
            maxProfit = profit;
            numbest = i; 
            bestset = include; 
        }

        if(promising(i, profit, weight)) {
            include[i+1] = true;        // incldue the next node
            knapsack(i+1, (profit + p[i+1]), (weight + w[i+1]));
            include[i+1] = false;       // don't include the next node
            knapsack(i+1, profit, weight);
        }
    }

    public static boolean promising (int i, int profit, int weight) {
        System.out.println("promising - "+ i);
        int j, k;
        j = k = 0;
        int totalWeight; 
        int bound; 

        if(weight >= capacity) {
            System.out.println("promising false - "+ i);
            return false;       // if we are over the weight limit, the node is not promising
        }
        else {
            j = i + 1; 
            bound = profit;
            totalWeight = weight;
            while ((j <= numItems) && (totalWeight+w[j] <= capacity))       // grab all possible items
            {
                totalWeight = totalWeight + w[j];
                bound = bound + p[j]; 
                j++;
            }
            k = j; 
            if (k <= numItems) {
                bound = (int) bound + (capacity - totalWeight) * (p[k]/w[k]); // recalculate the bound if k is not greater than number of Items 
            }
            System.out.println("promising trueish - "+ i);
            return bound > maxProfit; 
        }
    }

    public static void main (String [] args)
      throws IOException 
    {
        Scanner txtfile=new Scanner(new FileReader("knapsack.txt"));    // the items will be in the knapsack.txt file

        int itemW, itemP;

        try {
            capacity = txtfile.nextInt();   // first number in the text file is the capacity
            numItems = txtfile.nextInt();   // second number in the text file is the number of items in the file

            w = new int [numItems];
            p = new int [numItems];
            include = new boolean[numItems];
            bestset = new boolean[numItems];


            for (int i = 0; i < numItems; i++) {    // get the weights and the profit from the text file
                itemW = txtfile.nextInt();  
                itemP = txtfile.nextInt();
                w[i] = itemW; 
                p[i] = itemP; 
            }

            System.out.println("List of Items: ");
            //Print the items in the array after they are read in from the text file
            for(int i = 0; i < numItems; i++) {
                System.out.println("    ID: "+(i+1)+"  Weight: "+w[i]+"  Profit: "+p[i]);
            }

            System.out.println();
            System.out.println();
            
            numbest = 0;
            maxProfit = 0;
            knapsack(0, 0, 0);
            System.out.println("Calculations done");
            System.out.println("Maximum Profit: $"+maxProfit);
            System.out.println();
            System.out.print("Best Set = { ");
            for(int j = 0; j <= numbest; j++) {
                System.out.print(bestset[j]+" ");
            }
            System.out.print("} \n");


        } catch (Exception ex) {
            System.out.println("Error reading file.");
        }

        txtfile.close();
    }
}

File: Project 3/backtrack.cpp
/*
    Gabbi Forsythe
    12/12/2019
    CS351-01
    Dr. Chen

    
    PROJECT 3: Part 1 - 0/1 Knapsack (Backtracking Algorithm)
    Use the backtracking algorithm to find the solution to the 0/1 Knapsack problem
*/

#include <iostream>
#include <fstream>
#include <string>
using namespace std;

// global variables because that is what the book said to do on page 243
int n, W, numbest, maxprofit; 

// tells if the node is promising or not
bool promising (int i, int weight, int profit, int w [], int p []) { 
    int j, k, totalweight;
    float bound;

    if(weight >= W) {
        return false;   // if the weight exceeds the capacity of the knapsack
    }
    else {
        j = i+1;
        bound = profit; 
        totalweight = weight;
        while ((j <= n) && (totalweight +w[j] <= W)) {
            totalweight = totalweight + w[j];
            bound = bound + p[j];
            j ++;
        }
        k = j;
        if (k <= n) {
            bound = bound + (W - totalweight) * (p[k]/w[k]);
        }
        // cout << "Bound = " << bound; 
        return bound > maxprofit;       // if the bound is greater than the maximum profit, the node is promising
    }
}

void knapsack (int i, int profit, int weight, int* w, int* p, bool* include, bool* bestset)  {
    if ((weight <= W) && (profit > maxprofit))  {   // if it is the max profit so far 
        maxprofit = profit; 
        numbest = i;
        bestset = include; 
    }

    bool promise = promising(i, weight, profit, w, p); 

    if(promise == true ) {      // if the node is promising
        include[i+1] = true;        // include the node
        knapsack (i + 1, profit + p[i+1], weight + w[i+1], w , p, include, bestset);
        include[i+1] = false;        // don't include the node
        knapsack (i + 1, profit, weight, w , p, include, bestset);
    }
}

int main() { 

    ifstream File;
    File.open("knapsack.txt");

    File >> W;      // get the capacity from the file
    File >> n;      // get the number of items  

    // cout << W << " " << n << endl;   

    int w [n]; 
    int p [n];

    for(int i = 0; i < n; i++) {        // read the items in from the text file
        File >> w[i];
        File >> p[i];
        // cout << w[i] << " " << p[i] << endl;
    }

    cout << "List of Items: " << endl;
    //Print the items in the array after they are read in from the text file
    for(int i = 0; i < n; i++) {
        cout << "    ID: " << (i+1) << "  Weight: " << w[i] << "  Profit: " << p[i] << endl;
    }

    bool include [n+1];
    bool bestset [n+1];

    for (int g = 0; g < n+1; g++) {      // give bool arrays default values
        include[g] = false;
        bestset[g] = false; 
    }

    numbest = 0;
    maxprofit = 0; 
    knapsack(0, 0, 0, w, p, include, bestset);

    
    cout << "Max Profit: $" << maxprofit << endl; 
    cout << "Items included: "<< endl;
    for (int j = 1; j <= numbest; j++) {
        if (bestset[j] = true)
        cout << "    " << j << endl;
    }

    File.close();

    return 0;
}


/*
    KNAPSACK.TXT #1:
        30
        3
        5 50
        10 60
        20 140

    TEST RUN  #1: 
        thomas% a.out
        List of Items: 
            ID: 1  Weight: 5  Profit: 50
            ID: 2  Weight: 10  Profit: 60
            ID: 3  Weight: 20  Profit: 140
        Max Profit: $200
        Items included: 
            1
            2
        thomas%

    KNAPSACK.TXT #2:
        16
        4
        2 40
        5 30
        10 50
        5 10
    
    TEST RUN #2: 
        thomas% a.out
        List of Items: 
            ID: 1  Weight: 2  Profit: 40
            ID: 2  Weight: 5  Profit: 30
            ID: 3  Weight: 10  Profit: 50
            ID: 4  Weight: 5  Profit: 10
        Max Profit: $80
        Items included: 
            1
            3
        thomas% 
*/